### DISCUSSION TITLE
Does withCookies automatically prevent cookie tampering?

### DISCUSSION THREAD

**Discussion by tanin47** [playframework/playframework#12855] [Q&A]
I've played around with it, and it doesn't seem to automatically prevent cookie tampering. I wonder if that is the right understanding.

**Comment by tanin47** [comment]
It doesn't. Sessions does prevent tampering but the persistence and characteristics are not same as cookies... unfortunately.

**Reply by mkurz** [reply]
Hey, a bit late here, but what do you exactly mean with cookie tampering? I mean I know what in means in general, but what would you expect that `withCookies` should do for you automatically?

**Reply by tanin47** [reply]
No worries.

For context, I was surprised that we can edit a cookie through Chrome Inspect and PlayFramework would use that value without validating it. But I suppose that is expected.

I eventually switch to using `withSession` since it seems to encrypt the value using the application secret key.

The gap that I had is that I didn't realize a session can be persisted across browser session. Essentially, it's a normal cookie.

**Reply by mkurz** [reply]
> I eventually switch to using withSession since it seems to `encrypt` the value using the application secret key.

Be aware: The session cookie is signed using the secret, but is **NOT** encrypted. That "only" means the session cookie is signed with a secret key so the client can’t modify the cookie data (or it will be invalidated). See https://www.playframework.com/documentation/3.0.x/SettingsSession 

> Essentially, it's a normal cookie.

Yes, from the browser's perspective also the session is just a normal cookie. Depending on the `max-age` attribute ([you can configure yourself](https://github.com/playframework/playframework/blob/3.0.6/core/play/src/main/resources/reference.conf#L148-L151)) it will be persisted across browser session or not.
Actually, you can set the `secure`, `httpOnly`, `SameSite` and `Path` attributes of every cookie you set in Play in your response. So if you make a cookie `httpOnly` the browser will not be able to manipulate it anymore with javascript (again, this works not just with the session cookie but with every other cookie also)

**Reply by tanin47** [reply]
oops yes! I was gonna say it prevented tampering. I used a wrong word. Thank you for correcting me.

I had an old impression of PHP Session and thought it wouldn't be persisted across browser sessions. That was essentially the gap. After I understand it, Play's session is exactly what I want.
